app/multitenant/consul_pipe_router.go

```diff
@@ -0,0 +1,605 @@
+}
+
+func (c *consulPipe) toBytes() ([]byte, error) {
+	buf := bytes.Buffer{}
+	if err := json.NewEncoder(&buf).Encode(c); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func (c *consulPipe) fromBytes(bs []byte) error {
+	return json.NewDecoder(bytes.NewReader(bs)).Decode(c)
+}
+
+func (c *consulPipe) setEnd(e app.End, addr string) {
+	if e == app.UIEnd {
+		c.UIEnd = addr
+	} else {
+		c.ProbeEnd = addr
+	}
+}
+
+func (c *consulPipe) end(e app.End) string {
+	if e == app.UIEnd {
+		return c.UIEnd
+	}
+	return c.ProbeEnd
+}
+
+func (c *consulPipe) otherEnd(e app.End) string {
+	if e == app.UIEnd {
+		return c.ProbeEnd
+	}
+	return c.UIEnd
+}
+
+func (c *consulPipe) eitherEndFor(addr string) bool {
+	return c.end(app.UIEnd) == addr || c.end(app.ProbeEnd) == addr
+}
+
+func (c *consulPipe) incr(e app.End) int {
+	if e == app.UIEnd {
+		c.UIRef++
+		return c.UIRef
+	}
+	c.ProbeRef++
+	return c.ProbeRef
+}
+
+func (c *consulPipe) decr(e app.End) int {
+	if e == app.UIEnd {
+		c.UIRef--
+		return c.UIRef
+	}
+	c.ProbeRef--
+	return c.ProbeRef
+}
+
+type consulPipeRouter struct {
+	prefix    string
+	advertise string // Address of this pipe router to advertise in consul
+	client    *consul.Client
+	userIDer  UserIDer
+
+	mtx     sync.Mutex
+	cond    *sync.Cond
+	pipes   map[string]xfer.Pipe // Active pipes
+	bridges map[string]*bridgeConnection
+
+	// Used by Stop()
+	quit chan struct{}
+	wait sync.WaitGroup
+}
+
+// NewConsulPipeRouter returns a new consul based router
+func NewConsulPipeRouter(addr, prefix, inf string, userIDer UserIDer) (app.PipeRouter, error) {
+	advertise, err := network.GetFirstAddressOf(inf)
+	if err != nil {
+		return nil, err
+	}
+	client, err := consul.NewClient(&consul.Config{
+		Address: addr,
+		Scheme:  "http",
+	})
+	if err != nil {
+		return nil, err
+	}
+	pipeRouter := &consulPipeRouter{
+		prefix:    prefix,
+		advertise: advertise,
+		client:    client,
+		userIDer:  userIDer,
+
+		pipes:   map[string]xfer.Pipe{},
+		bridges: map[string]*bridgeConnection{},
+		quit:    make(chan struct{}),
+	}
+	pipeRouter.cond = sync.NewCond(&pipeRouter.mtx)
+	pipeRouter.wait.Add(1)
+	go pipeRouter.watchAll()
+	go pipeRouter.privateAPI()
+	return pipeRouter, nil
+}
+
+func (pr *consulPipeRouter) Stop() {
+	close(pr.quit)
+	pr.wait.Wait()
+}
+
+// watchAll listens to all pipe updates from consul.
+// This is effectively a distributed, consistent actor routine.
+// All state changes for this pipe router happen in this loop,
+// and all the methods are implemented as CAS's on consul, to
+// trigger an event in this loop.
+func (pr *consulPipeRouter) watchAll() {
+	var (
+		index = uint64(0)
+		kv    = pr.client.KV()
+	)
+	for {
+		select {
+		case <-pr.quit:
+			pr.wait.Done()
+			return
+		default:
+		}
+
+		kvps, meta, err := kv.List(pr.prefix, &consul.QueryOptions{
+			RequireConsistent: true,
+			WaitIndex:         index,
+			WaitTime:          longPollDuration,
+		})
```

Comment:
Will calling this in a loop miss updates while the callback is running? or is that handled by WaitIndex somehow?